src/types/utility-types/union.ts

```diff
@@ -35,7 +30,11 @@ export interface UnionOptions {
  * @internal
  * @hidden
  */
-export class Union extends BaseType<any, any, any> {
+export class Union<Types extends IAnyType[]> extends BaseType<
```

Comment:
I'll try my best. Hope this helps explain!

First, we need to know that there are tuple types (e.g., `["a", "b", "c"]`) and array types (e.g., `string[]`) in TypeScript. When we have this:

```ts
export function union<Types extends IAnyType[]>(...types: Types): IUnionType<Types>
```

That `Types extends IAnyType[]` could result in `Types` being either a tuple or array type, depending on how we call this function. When we do the following

```ts
types.union(types.literal("a"), types.literal("b"), types.literal("c"))
```

we'll find that `Types` is `[ISimpleType<"a">, ISimpleType<"b">, ISimpleType<"c">]`, and that's because TS is nice enough to narrow that `types` array to a tuple type (generally) because it knows statically that it is _exactly_ those three things. If we were to do the following

```ts
types.union(..."abc".split("").map(c => types.literal(c))
```

We'll find that `Types` is `ISimpleType<string>[]`. That's because `split` produced `string[]`, not `["a", "b", "c"]`. In theory, we could make `split` produce the right tuple type, but even then we wouldn't get what we want because:

```ts
const chars = ["a", "b", "c"] as const;
types.union(...chars.map(c => types.literal(c))
```

In that case, we get `ISimpleType<"a" | "b" | "c">[]` because `map` isn't typed to deal with (and I'm not sure we could make that work without higher-kinded types).

So that's how we got our tuple type, which is key to making the union type work the way we want it. The final bit is how do we pull out creation types and whatnot from that? Well, fortunately you can ask TypeScript to give you the type based on a certain index type for an array/tuple:

```ts
const a = ["a", "b", "c"] as const;
const b = a as string[];

// typeof a[number] == "a" | "b" | "c";
// typeof b[number] == string;
```

Makes sense, because if you're saying I have an array with exactly three items, the first being `"a"`, the second `"b"`, and the last `"c"`, if I index with some arbitrary number it has to be one of them, i.e., `"a" | "b" | "c"`. We could also do `typeof a[1]` to get `"b"`.

TS goes even further that if all parts of the union type share something in common, you can index that again. For example:

```ts
type Things = [{ prop: 1 }, { prop: boolean }];
type ThingsProp = Things["prop"];  // 1 | boolean
```

Using thius last bit of knowledge we can then construct our creation/snapshot/instance types:

```ts
// From our example above, `Types == [ISimpleType<"a">, ISimpleType<"b">, ISimpleType<"c">]`
// Given that, `Types[number]` will be `ISimpleType<"a"> | ISimpleType<"b"> |ISimpleType<"c">

type IUnionType<Types extends IAnyType[]> = ITypeUnion<
  Types[number]["CreationType"],   // "a" | "b" | "c"
  Types[number]["SnapshotType"],   // "a" | "b" | "c"
  Types[number]["TypeWithoutSTN"]  // "a" | "b" | "c"
>
```

From here, you may be able to see how this translates to the complex types like models and maps.

Reply:
Thanks man! That's really helpful. I appreciate it a lot.